midp/gfx.js

```diff
@@ -112,8 +112,35 @@ var currentlyFocusedTextEditor;
         return 1;
     };
 
+    var hasNewFrame = true;
+    var ctxToExecuteOnNewFrame;
+    function gotNewFrame(timestamp) {
+        if (ctxToExecuteOnNewFrame) {
+            var ctx = ctxToExecuteOnNewFrame;
+            ctxToExecuteOnNewFrame = null;
+            window.requestAnimationFrame(gotNewFrame);
+            ctx.execute();
+        } else {
+            hasNewFrame = true;
+        }
+    }
 
-    addUnimplementedNative("com/sun/midp/lcdui/DisplayDevice.refresh0.(IIIIII)V");
+    Native["com/sun/midp/lcdui/DisplayDevice.refresh0.(IIIIII)V"] = function() {
+        // This is a no-op: The foreground display gets drawn directly
+        // to the screen.
+        //
+        // However, we sometimes "draw" more frequently than the
+        // screen can update. By calling `requestAnimationFrame` here,
+        // we ensure that whatever we just drew will stick around for
+        // one frame
+        if (hasNewFrame) {
+            hasNewFrame = false;
+            window.requestAnimationFrame(gotNewFrame);
+        } else {
+            ctxToExecuteOnNewFrame = $.ctx;
+            $.pause(asyncImplStringAsync);
```

Comment:
I'm still trying to figure out some of the event queue/repaint stuff, but I'm starting to think synchronization with request animation frame should happen further up repaint call. Refresh0 is called at the very end of repaint whereas you'd typically want to start your drawing on the beginning of the request animation frame. I think this should happen in RepaintEventProducer in either process() or scheduleRepaint(). As far as I can tell on desktop we repaint too often(we can repaint under 3ms) whereas on a flame repaint takes 80ms-300ms.